features/375-static-cdn/design.md

```diff
@@ -0,0 +1,77 @@
+# Static Asset Hosting
+
+## Background
+Developer would want to host static content (e.g. single-page-application files) on Skygear and serve them through
+high performance CDN.
+
+## Objectives
+- Developers should be able to upload static content to be hosted at specific path.
+- Hosted static content should be served & cached with CDN if possible.
+- SPA style routing should be supported.
+
+## Design
+
+### Static Deployment
+
+A new type of deployment, `static`, would be supported. This deployment type
+supports configurating a deployment context (`context`) and fallback rewrite
+path (`fallback`).
+
+When `static` deployment type is used:
+- The deployment context would be persisted to object storage.
+- A request with matching path would be routing using static deployment routing
+  algorithm (described below).
+- If a file is matched, the gateway would proxy the request to the backing
+  object storage file.
+- If no file is matched, a generic 404 page would be returned.
+- The response would have appropiate cache control header, allowing maximum
+  cache duration of 24 hour.
+
+### Static Deployment Routing
+
+When a request matched a `static` deployment, only its path would be considered
+during routing. The request would be routed as followed:
+
+1. Let `U` be the request URI, let `P` be the routing path.
+2. If file `P` exists, return `P`.
+3. If file `Q` = `P/index.html` exists, return `Q`.
```

Comment:
I think it is equivalent to making `fallback = /index.html` by default? Why do we want to make this a special case? A generalization of `fallback` may be the `try_files` directive of nginx. How do we handle trailing slash?

Reply:
It's like using nginx directive `index index.html` and `try_files $uri <fallback>`. The step 3 is mainly for the convention of index file, and fallback path is an absolute path. Trailing slashes are ignored.

Reply:
Can we document the behavior in terms of nginx directives as in your comment? It would be helpful for some readers who are familiar with nginx.

So `fallback` must be an absolute path, right? Should we clarify this?

Reply:
Clarified routing behavior and configuration.